pkg/handlers/apk.go

```diff
@@ -0,0 +1,430 @@
+	dexOutput.Write(classOutput.Bytes())
+
+	// Stringify the classOutput value for case-insensitive keyword matching
+	classOutputLower := strings.ToLower(classOutput.String())
+
+	// Check if classOutput contains any of the default keywords
+	foundKeywords := make(map[string]struct{})
+	for keyword := range defaultKeywords {
+		if strings.Contains(classOutputLower, keyword) {
+			foundKeywords[keyword] = struct{}{} // Directly add to the map
+		}
+	}
+
+	// For each found keyword, create a keyword:value pair and append to dexOutput
+	var keyValuePairs bytes.Buffer
+	for str := range methodValues {
+		for keyword := range foundKeywords {
+			keyValuePairs.Reset()
+			keyValuePairs.WriteString(keyword + ":" + str + "\n")
+			dexOutput.Write(keyValuePairs.Bytes())
+		}
+	}
+}
+
+// processDexMethod iterates over a slice of methods, processes each method,
+// handles errors, and writes the output to dexOutput.
+func processDexMethod(ctx logContext.Context, dexReader *dextk.Reader, methods []dextk.MethodNode, classOutput *bytes.Buffer, methodValues map[string]struct{}) {
+	for _, method := range methods {
+		s, values, err := parseDexInstructions(dexReader, method)
+		if err != nil {
+			ctx.Logger().V(2).Info("failed to process dex method", "error", err)
+			continue
+		}
+		classOutput.Write(s.Bytes())
+		for val := range values {
+			methodValues[val] = struct{}{}
+		}
+	}
+}
+
+// parseDexInstructions processes a dex method and returns the string representation of the instruction
+func parseDexInstructions(r *dextk.Reader, m dextk.MethodNode) (*bytes.Buffer, map[string]struct{}, error) {
+	var s bytes.Buffer
+	values := make(map[string]struct{})
+
+	if m.CodeOff == 0 {
+		return &s, values, nil
+	}
+
+	c, err := r.ReadCodeAndParse(m.CodeOff)
+	if err != nil {
+		return &s, values, err
+	}
+
+	// Iterate over the instructions and extract the relevant values
+	for _, o := range c.Ops {
+		oStr := o.String()
+		// Filter out instructions that are not in our targetInstructionTypes
+		parsedVal := formatAndFilterInstruction(oStr)
+		if parsedVal == "" {
+			continue
+		}
+		// If instruction is a const-string, then store as in values map
+		// this is used when creating keyword:value pairs in processDexClass()
+		if strings.HasPrefix(oStr, stringInstructionType) {
+			values[parsedVal] = struct{}{}
+		}
+		// Write the parsedVal to the buffer
+		s.WriteString(parsedVal)
+	}
+	return &s, values, nil
+}
+
+// formatAndFilterInstruction looks for a match to our regex and returns it
+// Note: This is critical for ensuring secret + keyword are in close proximity.
+// If we expand the instructions we're looking at, this function will need to be updated.
+func formatAndFilterInstruction(line string) string {
+	for _, re := range reInstructions {
+		matches := re.FindStringSubmatch(line)
+		if len(matches) > 1 {
+			return matches[1]
+		}
+	}
+	return ""
+}
+
+func decodeXML(rdr io.ReadCloser, resTable *apkparser.ResourceTable) (io.Reader, error) {
```

Comment:
Great points. Please lmk if [224dab1](https://github.com/trufflesecurity/trufflehog/pull/3517/commits/224dab148ce7aabf53b7d1367d28efc60c787d3c) addresses this sufficiently.

Reply:
Using `bytes.Buffer` as the `io.Writer` for `xml.NewEncoder `could potentially lead to high memory usage if the buffer grows large. However, I think we can ignore that for now.

In the future, we might consider a custom writer that spills to disk if needed, but this can wait. I _think_ we might have a pkg that does that, but i'm not certain.